### ISSUE TITLE
Documentation on why error on training dataset differs during training and when applying the model to dataset.

### ISSUE THREAD

**Issue by AlexGrig** [catboost/catboost#443]
Hello,

I observe that the training metric e.g Recall do match between a standard output (or tensorboard) and explicit computations. There was a similar issue #310 which concerned validation metrics, but I observe discrepancy for the trainig metric.

Here is an example code:

```
import catboost
from catboost.datasets import titanic
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score

# data preparation
train_df, test_df = titanic()
train_df.fillna(-999, inplace=True)
test_df.fillna(-999, inplace=True)
X, y = train_df.drop('Survived', axis=1), train_df.Survived
categorical_features_indices = np.where(X.dtypes != np.float)[0]
X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.3, random_state=40, shuffle = True)

# training
params = { 'iterations': 15, 'eval_metric': 'Recall', 'logging_level': 'Silent', 'use_best_model': False, 'random_seed': 0, 'train_dir': './titanic_experiment'}
train_pool = catboost.Pool(X_train, y_train, cat_features=categorical_features_indices)
validate_pool = catboost.Pool(X_validation, y_validation, cat_features=categorical_features_indices)
model = catboost.CatBoostClassifier(**params)
model.fit(train_pool, eval_set=validate_pool, verbose=True)
```
The trainig output is (the same it shows in tensorboard):

```
Learning rate set to 0.5
0:	learn: 0.6565217	test: 0.7321429	best: 0.7321429 (0)	total: 26ms	remaining: 156ms
1:	learn: 0.4478261	test: 0.5357143	best: 0.7321429 (0)	total: 48.7ms	remaining: 122ms
2:	learn: 0.6565217	test: 0.7321429	best: 0.7321429 (0)	total: 83.7ms	remaining: 112ms
3:	learn: 0.5043478	test: 0.5714286	best: 0.7321429 (0)	total: 112ms	remaining: 83.7ms
4:	learn: 0.5260870	test: 0.5803571	best: 0.7321429 (0)	total: 135ms	remaining: 54ms
5:	learn: 0.5130435	test: 0.5714286	best: 0.7321429 (0)	total: 157ms	remaining: 26.1ms
6:	learn: 0.5565217	test: 0.6160714	best: 0.7321429 (0)	total: 183ms	remaining: 0us

bestTest = 0.7321428571
bestIteration = 0
```

Let's compute the recall by 2 ways: sklearn and catboost (outputs are in comments):

```
print(recall_score(y_train, model.predict(X_train))) # 0.8434782608695652
print(model.eval_metrics(train_pool, metrics=['Recall'])['Recall'][-1] ) # 0.8434782608695652

print(recall_score(y_validation, model.predict(X_validation))) # 0.6160714285714286
print(model.eval_metrics(validate_pool, metrics=['Recall'])['Recall'][-1]) # 0.6160714285714286
```
As you can see, the recall on validation data is the same as output from the model training. However, on the training data the recall does not match with the training output. In tensorboard the it shows the same as in training output. I suspect that happens not only for recall but for other metrics also.

Could you comment on that?

catboost version: 0.9.1.1
Operating System: macOS High Sierra
CPU: 2 GHz Intel Core i7

**Comment by annaveronika** [comment]
This is expected behavoir. During training we calculate numerical features based on categorical features in the way described here: https://tech.yandex.com/catboost/doc/dg/concepts/algorithm-main-stages_cat-to-numberic-docpage/ and here http://learningsys.org/nips17/assets/papers/paper_11.pdf

Namely, we perform a random permutation of the training dataset and for each example in the training dataset we compute average label value for the example with the same category value placed before the given one in the permutation and make it the new numerical feature value.

When we are making predictions, we are using the whole training dataset to compute these averages.
For an example in the training dataset during prediction the new numerical feature value will be equal to average label value of all training samples with this categorical feature value. 

Thus, new numerical feature values for the training dataset (not for the test dataset) are different during training and during prediction.

**Comment by AlexGrig** [comment]
Thank you for the explanation.
The issue #296 seems to be pretty much the same.

**Comment by annaveronika** [comment]
You are right, this is the second issue on this topic. We'll add better documentation for that.

**Comment by Rajat700** [comment]
Is the same behavior accepted if I use on_hot_max_size parameter and do one hot encoding of all the features. ?

**Comment by annaveronika** [comment]
@Rajat700 No, this should only happen if you are using target based statistics. With one-hot it should be the same. Do you encounter different behavoir?

**Comment by Rajat700** [comment]
No I checked with one hot then the training log error is exactly matching with the training error during prediction.
Thanks for the clarification

**Comment by david-waterworth** [comment]
I had the same question today - if the reported fit metrics don't match the loss calculated by predict, which metric is used for cross validation?

**Comment by annaveronika** [comment]
> I had the same question today - if the reported fit metrics don't match the loss calculated by predict, which metric is used for cross validation?

That is a great question! We didn't write this in docs also, we'll add this. Yes, with cv it is the same as with train - train dataset error is calculated using the dynamically built features. Note that test error uses the features built on the whole train, so test error is the same.

**Comment by annaveronika** [comment]
> No I checked with one hot then the training log error is exactly matching with the training error during prediction.
> Thanks for the clarification
> 
> But then my question is if I use target based statistics then to check overfitting which error should I use? Should I use the training log error or the one which I am getting by using the complete training set for prediction to find out whether my model is overfitted or not ?

You should use test error that is reported during training. It is exactly the same as one you get when doing predictions.
During training you see error on train dataset and error on test dataset. You need error on test dataset to check overfitting. This one is same during training and during applying.
Error on train dataset during training might be different, because train dataset has dynamically built features that are different during training and during applying.

**Comment by annaveronika** [comment]
This is published now:
https://catboost.ai/docs/concepts/faq.html#applying-the-model-to-train-dataset
